### PR TITLE
NMS-10226: Two BridgePort Node - Topology Mismatch

### PR DESCRIPTION
**Pull request by rssntn67**
* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-10226

### CODE REVIEW

File: features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java [hunk 1]
```diff
@@ -148,106 +149,120 @@ private BridgeSimpleConnection(BridgeForwardingTable xBridge,
...
+            List<BridgePort> ports = BridgeSimpleConnection.condition3(commonlearnedmacs, m_xBridge, m_yBridge);
+            if (ports.size() == 2) {
+                m_xyPort = ports.get(0);
+                m_yxPort = ports.get(1);
+                return;
+            }
+        }
+
+        if (m_xyPort == null && m_yxPort == null) {
+            List<BridgePort> ports=BridgeSimpleConnection.condition4(m_xBridge, m_yBridge);
```
Comment (christianpape) [inline features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java:218]:
```suggestion
            List<BridgePort> ports = BridgeSimpleConnection.condition4(m_xBridge, m_yBridge);
```

File: features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java [hunk 2]
```diff
@@ -148,106 +149,120 @@ private BridgeSimpleConnection(BridgeForwardingTable xBridge,
...
+        if (m_xyPort == null && m_yxPort == null) {
+            List<BridgePort> ports = BridgeSimpleConnection.condition3(commonlearnedmacs, m_xBridge, m_yBridge);
+            if (ports.size() == 2) {
+                m_xyPort = ports.get(0);
+                m_yxPort = ports.get(1);
+                return;
+            }
+        }
+
+        if (m_xyPort == null && m_yxPort == null) {
```
Comment (christianpape) [inline features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java:217]: Sam condition like line #208. Can be combined together, right?

File: features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java [hunk 3]
```diff
@@ -148,106 +149,120 @@ private BridgeSimpleConnection(BridgeForwardingTable xBridge,
...
+                m_xyPort = ports.get(0);
+                m_yxPort = ports.get(1);
+                return;
+            }
+        }
+        throw new BridgeTopologyException("findSimpleConnection: no simple connection found", m_xBridge);
+    }
+
+    private static List<BridgePort> condition4(BridgeForwardingTable bridgexFt, BridgeForwardingTable bridgeyFt
+    ) {
```
Comment (christianpape) [inline features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java:229]:
```suggestion
    private static List<BridgePort> condition4(BridgeForwardingTable bridgexFt, BridgeForwardingTable bridgeyFt) {
```

File: features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java [hunk 4]
```diff
@@ -148,106 +149,120 @@ private BridgeSimpleConnection(BridgeForwardingTable xBridge,
...
+                for (String mac : bpwm.getMacs()) {
+                    if (commonSegmentMacAddress.contains(mac)) {
+                        continue NEXT;
+                    }
+                }
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("condition4: simple connection found [{}] -> [{}]", bridgexPort.printTopology(), bpwm.getPort().printTopology());
+                }
+                bbports.add(0,bridgexPort);
+                bbports.add(1,bpwm.getPort());
```
Comment (christianpape) [inline features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java:254]:
```suggestion
                bbports.add(0, bridgexPort);
                bbports.add(1, bpwm.getPort());
```

File: features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java [hunk 5]
```diff
@@ -148,106 +149,120 @@ private BridgeSimpleConnection(BridgeForwardingTable xBridge,
...
+                return bbports;
+            }
+        }
+        LOG.warn("condition4: no simple connection found [{}] -> [{}]", bridgexFt.getNodeId(),bridgeyFt.getNodeId());
```
Comment (christianpape) [inline features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java:258]:
```suggestion
        LOG.warn("condition4: no simple connection found [{}] -> [{}]", bridgexFt.getNodeId(), bridgeyFt.getNodeId());
```

File: features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java [hunk 6]
```diff
@@ -273,7 +290,7 @@ private static List<BridgePort> condition3(Set<String> commonlearnedmacs,
         BridgePort yp2=null;
         BridgePort xp1=null;
         BridgePort xp2=null;
```
Comment (christianpape) [inline features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java:292]:
```suggestion
        BridgePort yp2 = null;
        BridgePort xp1 = null;
        BridgePort xp2 = null;
```

File: features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java [hunk 7]
```diff
@@ -273,7 +290,7 @@ private static List<BridgePort> condition3(Set<String> commonlearnedmacs,
...
-        List<BridgePort> bbports = new ArrayList<BridgePort>(2);
+        List<BridgePort> bbports = new ArrayList<>(2);
         for (String mac: commonlearnedmacs) {
             if (mac1 == null) {
                 mac1=mac;
```
Comment (christianpape) [inline features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java:296]:
```suggestion
                mac1 = mac;
```

File: features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java [hunk 8]
```diff
@@ -393,17 +410,24 @@ private static List<BridgePort> condition3(Set<String> commonlearnedmacs,
         	bbports.add(1, yp1);
         	return bbports;
         }
-        throw new BridgeTopologyException("condition3: bridge: no simple connection", 
-                                          bridgeyFt);
+        LOG.warn("condition3: no simple connection found [{}] -> [{}]", bridgexFt.getNodeId(),bridgeyFt.getNodeId());
```
Comment (christianpape) [inline features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java:413]:
```suggestion
        LOG.warn("condition3: no simple connection found [{}] -> [{}]", bridgexFt.getNodeId(), bridgeyFt.getNodeId());
```

File: features/enlinkd/tests/src/test/java/org/opennms/netmgt/enlinkd/BroadcastDomainTest.java [hunk 9]
```diff
@@ -2142,5 +2142,22 @@ public void testDuplicatedMac() throws BridgeTopologyException {
...
+        TwoUnresolvableTopology topology = new TwoUnresolvableTopology();
+        BroadcastDomain domain = new BroadcastDomain();
+        Bridge.create(domain,topology.nodeIdA);
+        Bridge.create(domain,topology.nodeIdB);
+        setBridgeElements(domain,topology.elemlist);
+        
+        DiscoveryBridgeTopology ndbt= new DiscoveryBridgeTopology(domain);
+        
+        ndbt.addUpdatedBFT((topology.nodeIdA),topology.bftA);
+        ndbt.addUpdatedBFT((topology.nodeIdB),topology.bftB);
```
Comment (christianpape) [inline features/enlinkd/tests/src/test/java/org/opennms/netmgt/enlinkd/BroadcastDomainTest.java:2157]:
```suggestion
        Bridge.create(domain, topology.nodeIdA);
        Bridge.create(domain, topology.nodeIdB);
        setBridgeElements(domain, topology.elemlist);
        
        DiscoveryBridgeTopology ndbt = new DiscoveryBridgeTopology(domain);
        
        ndbt.addUpdatedBFT((topology.nodeIdA), topology.bftA);
        ndbt.addUpdatedBFT((topology.nodeIdB), topology.bftB);
```

File: features/enlinkd/tests/src/test/java/org/opennms/netmgt/enlinkd/EnLinkdTestHelper.java [hunk 10]
```diff
@@ -183,6 +183,57 @@ public static <T> boolean checkLinks(Iterable<T> iterable, Predicate<T>... match
...
+        String maca="00e0b1bb39ba";
+        String macb="00e0b1bba38b";
+        String macc="0012cf5d318c";
+        String macd="00e0b1bbac4d";
+        
+        OnmsNode nodeA = new OnmsNode();
+        OnmsNode nodeB = new OnmsNode();
+        
+        BridgeElement bridgeElementA=new BridgeElement();
+        BridgeElement bridgeElementB=new BridgeElement();
```
Comment (christianpape) [inline features/enlinkd/tests/src/test/java/org/opennms/netmgt/enlinkd/EnLinkdTestHelper.java:205]:
```suggestion
        String basebridgeaddrA = "00e0b1bb39aa";
        String basebridgeaddrB = "00e0b1bb39bb";

        String maca = "00e0b1bb39ba";
        String macb = "00e0b1bba38b";
        String macc = "0012cf5d318c";
        String macd = "00e0b1bbac4d";
        
        OnmsNode nodeA = new OnmsNode();
        OnmsNode nodeB = new OnmsNode();
        
        BridgeElement bridgeElementA = new BridgeElement();
        BridgeElement bridgeElementB = new BridgeElement();
```

File: features/enlinkd/tests/src/test/java/org/opennms/netmgt/enlinkd/EnLinkdTestHelper.java [hunk 11]
```diff
@@ -183,6 +183,57 @@ public static <T> boolean checkLinks(Iterable<T> iterable, Predicate<T>... match
...
+            bftB.add(addBridgeForwardingTableEntry(nodeB, 10, 11010, macd, 1));
+            bftB.add(addBridgeForwardingTableEntry(nodeB, 90, 11090, maca, 1));
+            bftB.add(addBridgeForwardingTableEntry(nodeB, 90, 11090, macb, 1));
+
+        
+        }
+         
+        
+
+    }
```
Comment (christianpape) [inline features/enlinkd/tests/src/test/java/org/opennms/netmgt/enlinkd/EnLinkdTestHelper.java:236]:
```suggestion
            bftB.add(addBridgeForwardingTableEntry(nodeB, 90, 11090, macb, 1));
        }
    }
```

**Review by christianpape — CHANGES_REQUESTED** [review]
Only minor things to fix.

File: features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java (hunk 2 above)
Comment (rssntn67) [reply features/enlinkd/service/api/src/main/java/org/opennms/netmgt/enlinkd/service/api/BridgeSimpleConnection.java]: commit c0b86a54b8980714d18ad0ea2348ff0eda3da5ef
Author: Antonio <rssntn67@yahoo.it>
Date:   Mon May 23 22:34:52 2022 +0200

    Two BridgePort Node - Topology Mismatch

    According to pape Christian put together
    See convertion on PR 4738

**Review by christianpape — APPROVED** [review]
LGTM.

**Comment by fooker** [comment]
@rssntn67 please prefix your commits with the issue number to make them trackable. You can fix this for this PR by choosing squash-merge while merging.

**Review by fooker — APPROVED** [review]

**Review by christianpape — APPROVED** [review]
LGTM.

**Pull request merged**